src/fsharp/vs/IncrementalBuild.fs

```diff
@@ -504,197 +505,235 @@ module internal IncrementalBuild =
         
     type Target = Target of INode * int option
 
+    module List =
+        let rec private foldImpl (folder, pending : 'T list, state : 'State) =
+            async {
+            match pending with
+            | [] -> return state
+            | el :: pending ->
+                let! state = folder state el
+                return! foldImpl (folder, pending, state)
+            }
+        
+        let foldAsync (folder : 'State -> 'T -> Async<'State>) (state : 'State) (list : 'T list) : Async<'State> =
+            foldImpl (folder, list, state)
+
     /// Visit each executable action necessary to evaluate the given output (with an optional slot in a
     /// vector output). Call actionFunc with the given accumulator.
-    let ForeachAction ctok (Target(output, optSlot)) bt (actionFunc:Action -> 'T -> 'T) (acc:'T) =
-        let seen = System.Collections.Concurrent.ConcurrentDictionary<Id,bool>()
-        let isSeen id = 
-            if seen.ContainsKey id then true
-            else 
-                seen.[id] <- true
-                false
-                 
-        let shouldEvaluate(bt,currentsig:InputSignature,id) =
-            if currentsig.IsEvaluated then 
-                currentsig <> Signature(bt,id)
-            else false
+    let ForeachAction ctok (Target(output, optSlot)) bt (actionFunc:Action -> 'T -> Async<'T>) (acc:'T) : Async<'T> =
+        async {
```

Comment:
I mean changing

```
let f x = 
    let a = b
    let a = b
    let a = b
    let a = b
    c
```
to
```
let f x = 
  async {
    let a = b
    let a = b
    let a = b
    let a = b
    return c
  }
```
i.e inserting the monad using two-space and not changing most of the function (at least initially while getting the basic PR into shape).  It's just really useful to see the minimized diff and be able to comment on it directly on github